### PR TITLE
Add flow for displaying achieved medals

### LINKED ISSUE #18880: Add client-server loop for awarding medals
We already have a (very dated) medal award implementation, so the first step should be to get the communication working so the server can let us know when new medals are achieved. Should likely use SignalR, potentially as part of the new `BeatmapHub` i'm working on.

This time around, we hope to make medals awarded outside of the results screen, so we can award them for things that aren't necessarily related to a score being submitted.

### PR DESCRIPTION
**Pull request by bdach**
- Closes https://github.com/ppy/osu/issues/18880
- [x] Depends on https://github.com/ppy/osu-web/pull/11018 for display of descriptions

This is tested to work, but PRing as draft to confront direction.

https://github.com/ppy/osu/assets/20418176/4623da9a-22b4-45d3-8037-3787004db8f3

https://github.com/ppy/osu/assets/20418176/55fe7a6f-4f97-4bbe-9d21-fcac183db5d8

- Uses old design for the time being.
- Handles queueing of displaying multiple medals awarded in quick succession one after another.
- Respects overlay activation mode (so will not display mid-gameplay, for instance). Medals that have not been shown due to incorrect activation mode will display at the first available chance.

Reading commit-by-commit is recommended, the diff overexaggerates the extent of changes. Most of the diffstat comes from the `MedalOverlay` rename - it was really underbaked for the expectations I wanted to exercise from it as outlined above.

Tests are provided, but tests are tests. There are two ways to test on something resembling a real environment. The simpler version is as follows:

1. Set up a local [`osu-web` instance](https://github.com/ppy/osu/wiki/Testing-web-server-full-stack-with-osu!) etc.
2. Launch web, launch spectator server, launch client. Log in as a user in client.
3. Using `docker compose run --rm php mysql` in web directory, find the ID of user you logged onto in step (2), and an ID of a random achievement you like. (They're not gonna resemble real achievements, if it's what the web seeder generated for you.)
4. Use `docker compose run --rm php artisan tinker` in web directory, run:

    ```php
    UserAchievement::unlock(User::find(156), Achievement::find(191))
    ```

    IDs will obviously likely differ.
6. If you want to repeat step (4), delete the appropriate `osu_user_achievements` rows from the db.

You can also go the whole 9 yards and set things up such that a local `osu-web` instance and a local `osu-queue-score-statistics` instance talk to each other (which I did), but the steps to do so are long and involved so I'm not gonna bother.

This is tested to work, but I'm not sure this is the best way to make this happen. It is for sure the simplest, architecturally speaking, since it uses pieces we already have, but now we have yet another place with weird duality (score statistics updates are signalled by `osu-queue-score-processor` via redis, but now medals are done via hooking into existing web push notification flows).

When testing, ensure that you have push notifications enabled on your user on the web instance, since they can be disabled in there. Not sure if we will want to disable the possibility of disabling them via web or whatever.

### CODE REVIEW

**Comment by peppy** [comment]
Taking this out of draft as I think this direction is correct.

File: osu.Game/Properties/AssemblyInfo.cs [hunk 1]
```diff
@@ -11,3 +11,6 @@
 [assembly: InternalsVisibleTo("osu.Game.Tests.Dynamic")]
 [assembly: InternalsVisibleTo("osu.Game.Tests.iOS")]
 [assembly: InternalsVisibleTo("osu.Game.Tests.Android")]
+
+// intended for Moq usage
+[assembly: InternalsVisibleTo("DynamicProxyGenAssembly2")]
```
Comment (peppy) [inline osu.Game/Properties/AssemblyInfo.cs:16]: what's this one for?

Comment (bdach) [reply osu.Game/Properties/AssemblyInfo.cs:16]: It's for the test. The test uses a mock `IOverlayManager`, which is an internal interface.

Alternative would be making it public I guess?

Comment (peppy) [reply osu.Game/Properties/AssemblyInfo.cs:16]: It makes sense after googling what `DynamicProxyGenAssembly2` is so I guess it's okay.

**Review by peppy — APPROVED** [review]

**Pull request merged**